### PR TITLE
[Feature] Directory tree and multiple files support

### PR DESCRIPTION
**Pull request by BJordi**
## Summary

- Add directory tree to the code display.
- Refactor GitHub query and recipe interface.
- Refactor download feature.
- Change styles in order to fit the new design.

## Screenshot

002-directory-tree

### CODE REVIEW

File: app/src/app/screens/Dashboard/screens/Detail/index.tsx [hunk 1]
```diff
@@ -7,21 +7,31 @@ import { saveAs } from 'file-saver';
...
     // eslint-disable-next-line @typescript-eslint/no-unused-vars
-    const { title, css, config, ...files } = recipeCode;
-
+    const { title, config, source, ...files } = recipeCode;
     updateDownloadsCounter(tech, recipe);
-
     const zip = new JSZip();
+    // TODO: Remove TreeEntry type once the parseTree has type checking
+    source?.entries.forEach((entry: TreeEntry) => zipEntry(zip, entry));
     Object.values(files).forEach(file => {
```
Comment (pabloferro) [inline app/src/app/screens/Dashboard/screens/Detail/index.tsx:35]: This should be removed right? We have a double zip here

File: app/src/utils/recipes.ts [hunk 2]
```diff
@@ -10,6 +10,41 @@ const getRecipeByFileName = (recipe: GitHubResult, fileName: string) => {
     : null;
 };
 
+const parseAndSortTree = (entries: any) => {
+  const newSource = entries.map((entry: any) =>
```
Comment (pabloferro) [inline app/src/utils/recipes.ts:14]: We'll need to these `any`, probably in another PR.

File: app/src/utils/recipes.ts [hunk 3]
```diff
@@ -10,6 +10,41 @@ const getRecipeByFileName = (recipe: GitHubResult, fileName: string) => {
...
+        }
+  );
+  return newSource.sort((entryA: TreeEntry, entryB: TreeEntry) => entryA.type < entryB.type);
+};
+
+const filterBlobEntries = (entries: any) =>
+  entries.filter((entry: TreeEntry) => {
+    if (entry.type === FileTypes.tree) {
+      filterBlobEntries(entry.entries);
+    } else if (entry.name === 'cookbook.json') {
```
Comment (pabloferro) [inline app/src/utils/recipes.ts:36]: I think `cookbook.json` is repeated in this file, extract a constant.

File: app/package.json [hunk 4]
```diff
@@ -26,8 +26,10 @@
     "i18next": "^13.1.5",
     "jszip": "^3.4.0",
     "lodash": "^4.17.15",
+    "lodash.flatten": "^4.4.0",
     "moment": "^2.24.0",
     "node-sass": "^4.13.1",
+    "npm": "^6.14.5",
```
[the remark this answers is not in the document]
Comment (mcallegari10) [reply app/package.json:32]: I think not 😝

File: app/src/app/components/Tree/components/TreeFile/index.tsx [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+function TreeFile({
+  name,
+  src = '',
+  isBinary,
+  id,
+  activeId,
+  handleSelect,
+  paddingLeft = defaultPaddingLeft
+}: Props) {
+  const padding = { '--padding-left': `${paddingLeft}px` } as React.CSSProperties;
```
[the remark this answers is not in the document]
Comment (mcallegari10) [reply app/src/app/components/Tree/components/TreeFile/index.tsx:24]: It's a CSS Variable 😄

File: app/src/app/screens/Dashboard/screens/Detail/layout.tsx [hunk 6]
```diff
@@ -16,36 +16,33 @@ interface Props {
...
+      <div className={styles.upperSection}>
+        <div className="row space-between m-bottom-4 middle">
+          <h1 className="title">{title}</h1>
+          <a
+            href="https://github.com/Wolox/frontend-cookbook"
+            className={`row midle ${styles.githubLink}`}
+            target="_blank"
+            rel="noopener noreferrer"
+          >
+            by Tincho_otaku
```
Comment (mcallegari10) [inline app/src/app/screens/Dashboard/screens/Detail/layout.tsx:34]: Delete this please 😆  and add a TODO to add the Github user of the creator

File: app/src/app/screens/Dashboard/screens/Detail/layout.tsx [hunk 7]
```diff
@@ -16,36 +16,33 @@ interface Props {
...
+            <label htmlFor="chkButton" />
+          </div>
         </div>
       </div>
-      {isCodeVisible ? (
-        <Code html={html.content} scss={scss.content} onDownload={onDownload} />
-      ) : (
-        <Settings />
-      )}
+      {isCodeVisible ? <Code title={title} source={recipe.source} onDownload={onDownload} /> : <Settings />}
```
Comment (mcallegari10) [inline app/src/app/screens/Dashboard/screens/Detail/layout.tsx:45]: For now, there will be no `Settings` tab. Please delete the ternary or make the toggle not possible

File: app/src/app/screens/Dashboard/screens/Detail/layout.tsx (hunk 6 above)
Comment (pabloferro) [reply app/src/app/screens/Dashboard/screens/Detail/layout.tsx:34]: ⚰️

File: app/src/app/components/Tree/components/TreeFile/index.tsx [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  const lang = name.split('.').pop() || name;
+  return (
+    <button
+      type="button"
+      className={cn(styles['tree-item'], { [styles['tree-item--active']]: id === activeId })}
+      style={padding} // eslint-disable-line react/forbid-dom-props
+      onClick={() => handleSelect({ name, src, lang, isBinary, id })}
+    >
+      - {name.split('/').pop() || name}
```
[the remark this answers is not in the document]
Comment (BJordi) [reply app/src/app/components/Tree/components/TreeFile/index.tsx:33]: Yes, they do.

File: app/src/app/screens/Dashboard/screens/Detail/components/Code/index.tsx [hunk 9]
```diff
@@ -1,57 +1,59 @@
...
+    setFileData(data);
+  };
 
   return (
-    <div className={styles.codeContainer}>
-      <div className="row space-between m-bottom-4">
-        <div>
+    <div className={styles.filesStructureContainer}>
+      <div className={styles.header}>
+        <div className={styles['download-container']}>
```
Comment (pabloferro) [inline app/src/app/screens/Dashboard/screens/Detail/components/Code/index.tsx:33]:
```suggestion
        <div className={styles.downloadContainer}>
```

File: app/src/app/screens/Dashboard/screens/Detail/components/Code/index.tsx [hunk 10]
```diff
@@ -1,57 +1,59 @@
...
           </button>
         </div>
-        <button
-          className={`${styles.downloadButton} ${styles.codeTypeButton}`}
-          type="button"
-          onClick={onDownload}
-        >
-          Download code
-        </button>
+        <h3 className={styles['file-name']}>{fileData.name}</h3>
```
Comment (pabloferro) [inline app/src/app/screens/Dashboard/screens/Detail/components/Code/index.tsx:42]:
```suggestion
        <h3 className={styles.fileName}>{fileData.name}</h3>
```

File: app/src/app/screens/Dashboard/screens/Detail/layout.tsx [hunk 11]
```diff
@@ -16,36 +15,26 @@ interface Props {
...
-      </div>
-      <div className={styles.detailsRecipe}>
-        <RecipePreview recipe={recipe} />
-        <div className={styles.settingsCheckbox}>
-          <input type="checkbox" id="chkButton" onClick={handleClick} />
-          <label htmlFor="chkButton" />
+      <div className={styles.upperSection}>
+        <div className="row space-between m-bottom-4 middle">
+          <h1 className="title">{title}</h1>
+          <a
```
Comment (pabloferro) [inline app/src/app/screens/Dashboard/screens/Detail/layout.tsx:23]: Remove this entire element, I've added a card in backlog.

File: app/src/utils/queries.ts [hunk 12]
```diff
@@ -29,17 +29,65 @@ const queryBuilder = (
...
+                      name
+                      object {
+                        ... on Blob {
+                          id
+                          isBinary
+                          text
+                        }
+                        ... on Tree {
+                          entries {
+                            name
```
Comment (pabloferro) [inline app/src/utils/queries.ts:76]: Add object with blob or remove this level.
If you receive an entry without object wouldn't it break?

**Pull request merged**